File: app/src/main/java/com/kaappi/studio/viewmodel/FileBrowserViewModel.kt
package com.kaappi.studio.viewmodel

import androidx.lifecycle.ViewModel
import com.kaappi.studio.data.FileRepository
import com.kaappi.studio.data.FileRepositoryException
import com.kaappi.studio.data.SchemeFileNames
import com.kaappi.studio.domain.SchemeFile
import java.io.File
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Every operation is a `suspend` function: the repository does its disk I/O
 * on `Dispatchers.IO`, so UI callers launch these from the main thread
 * without ever blocking it (issue #12). Failures propagate as exceptions to
 * the caller, which owns the snackbar.
 */
class FileBrowserViewModel(private val fileRepository: FileRepository) : ViewModel() {

    private val _files = MutableStateFlow<List<SchemeFile>>(emptyList())
    val files: StateFlow<List<SchemeFile>> = _files.asStateFlow()

    suspend fun refresh() {
        _files.value = fileRepository.listFiles()
    }

    suspend fun saveFile(name: String, content: String): SchemeFile =
        fileRepository.writeFile(name, content).also { refresh() }

    /**
     * True when a file with this name (sanitized, base name) already exists on
     * disk. Used by the new-file flow so it can confirm before overwriting
     * (issue #10). Throws [IllegalArgumentException] for invalid names.
     */
    suspend fun fileExists(name: String): Boolean {
        val base = SchemeFileNames.sanitize(name)
        return fileRepository.listFiles().any { it.name == base }
    }

    /**
     * Deletes the file at [path]. The list is refreshed either way so it shows
     * what is actually on disk. Throws [FileRepositoryException] when nothing
     * was deleted (missing file, permission failure), matching the
     * repository's never-fabricate-success contract: callers must not reset
     * editor state for a file that still exists (issue #15).
     */
    suspend fun deleteFile(path: String) {
        val deleted = fileRepository.deleteFile(path)
        refresh()
        if (!deleted) throw FileRepositoryException("Cannot delete ${File(path).name}")
    }

    suspend fun renameFile(oldPath: String, newName: String): SchemeFile? =
        fileRepository.renameFile(oldPath, newName).also { refresh() }

    /** The file's contents; [SchemeFile] itself carries none (issue #12). */
    suspend fun readFile(path: String): String = fileRepository.readFile(path)
}

File: iosApp/KaappiStudio/ViewModels/FileBrowserViewModel.swift
import Foundation
@preconcurrency import shared

enum FileBrowserError: LocalizedError {
    case deleteFailed(path: String)

    var errorDescription: String? {
        switch self {
        case .deleteFailed(let path):
            return "Could not delete \((path as NSString).lastPathComponent)."
        }
    }
}

/// Main-actor front for the shared Kotlin `FileRepository`.
///
/// File names, sanitization, atomic writes, the never-fabricate-success rules
/// and the schemes directory layout all live in `shared` (see the `expect`
/// KDoc in `FileRepository.kt`); this class only adapts the suspend API to
/// Swift concurrency and keeps the published list current. Kotlin exceptions
/// declared with `@Throws` arrive here as `NSError`s whose
/// `localizedDescription` is the Kotlin message.
@MainActor
class FileBrowserViewModel: ObservableObject {
    @Published var files: [SchemeFile] = []

    private let repository: FileRepository

    init(repository: FileRepository = FileRepository()) {
        self.repository = repository
    }

    func refresh() async {
        // listFiles never throws for I/O (an unreadable directory lists as
        // empty — see the expect KDoc); the only error that can cross the
        // bridge is cancellation, and a cancelled refresh must leave the
        // list as it was rather than blank it.
        guard let listed = try? await repository.listFiles() else { return }
        files = listed
    }

    /// Throws on invalid names or write failures — it never fabricates success.
    @discardableResult
    func saveFile(name: String, content: String) async throws -> SchemeFile {
        let file = try await repository.writeFile(name: name, content: content)
        await refresh()
        return file
    }

    /// True when a file with this (sanitized) base name already exists, so the
    /// new-file flow can confirm before overwriting. Checks the disk, not the
    /// in-memory list: this view model instance can be separate from the one
    /// other views save through, so `files` may be stale.
    func fileExists(name: String) async throws -> Bool {
        let base = try SchemeFileNames.shared.sanitize(name: name)
        return try await repository.listFiles().contains { $0.name == base }
    }

    /// Deletes the file at `path`. The list is refreshed either way so it shows
    /// what is actually on disk; throws when nothing was deleted, matching the
    /// Android view model.
    func deleteFile(path: String) async throws {
        let deleted = try await repository.deleteFile(path: path).boolValue
        await refresh()
        if !deleted { throw FileBrowserError.deleteFailed(path: path) }
    }

    /// Throws instead of returning "" so a failed read can never lead to the
    /// real content being overwritten with empty text.
    func readFile(path: String) async throws -> String {
        try await repository.readFile(path: path)
    }
}
